<?xml version="1.0"?>
 
<robot name="dobot" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:property name="config_file" value="/home/tp03/anro_ws/src/urdf_tutorial/urdf_variables/variables.yaml" />
  <xacro:property name="config_dict" value="${xacro.load_yaml('variables.yaml')}"/>

  
  <xacro:property name="link0_heigth" value="${config_dict['link0']['length']}" />
  <xacro:property name="link1_heigth" value="${config_dict['link1']['length']}" />
  <xacro:property name="link2_length" value="${config_dict['link2']['length']}" />
  <xacro:property name="link3_length" value="${config_dict['link3']['length']}"/>
  <xacro:property name="link4_length" value="${config_dict['link4']['length']}"/>
  
  
  <xacro:property name="link0_width" value="0.15" />
  <xacro:property name="link1_width" value="0.1" />
  <xacro:property name="link2_width" value="0.08" />
  <xacro:property name="link3_width" value="0.08"/>
  <xacro:property name="link4_width" value="0.08"/>
  
  <material name="red">
    <color rgba="0.9 0.0 0.0 1.0"/>
  </material>
  
  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>
  
  <material name="green">
    <color rgba="0 0.8 0 1"/>
  </material>
   
  <link name="base_link">
  </link>

  <link name="link_0">
    <visual>
      <origin xyz="0 0 ${link0_heigth/2}" rpy="0 0 0"/>
      <geometry>
        <box size="${link0_width} ${link0_width} ${link0_heigth}"/>
      </geometry>
      <material name="red"/>
    </visual>
  </link>
  
  <link name="plack">
    <visual>
     <geometry>
      <box size="0.1 0.001 0.04"/>
     </geometry>
     <material name="blue"/>
    </visual>
  </link>
  
  <joint name="joint_one" type="revolute">
   <origin xyz="0 0 ${link0_heigth}" rpy="0 0 0"/>
   <parent link="link_0"/>
   <child link="link_1"/>
   <axis xyz="0 0 1"/>
   <limit effort="1000.0" lower="1.57" upper="-1.57" velocity="0.5"/>
  </joint>
  
  <link name="link_1">
    <visual>
     <origin xyz="0 0 ${link1_heigth/2}" rpy="0 0 0"/>
     <geometry>
      <box size="${link1_width} ${link1_width} ${link1_heigth}"/>
     </geometry>
     <material name="green"/>
    </visual>
  </link>
  
  <joint name="joint_two" type="revolute">
   <parent link="link_1"/>
   <child link="link_2"/>
   <axis xyz="1 0 0"/>
   <origin xyz="0 0 ${link1_heigth-0.01}" rpy="0 0 0"/>
   <limit effort="1000.0" lower="0" upper="-1.48" velocity="0.5"/>
  </joint>
  
  <link name="link_2">
   <visual>
    <origin xyz="0 0 ${link2_length/2}" rpy="0 0 0"/>
    <geometry>
     <cylinder length="${link2_length}" radius="${link2_width/2}"/>
    </geometry>
    <material name="red"/>
   </visual>
  </link>
  
  <joint name="joint_three" type="revolute">
   <parent link="link_2"/>
   <child link="link_3"/>
   <axis xyz="1 0 0"/>
   <origin xyz="0 0 ${link2_length}" rpy="0 0 0"/>
   <limit effort="1000.0" lower="-0.654" upper="-2.413" velocity="0.5"/>
  </joint>
  
  <link name="link_3">
   <visual>
    <origin xyz="0 0 ${link3_length/2}" rpy="0 0 0"/>
    <geometry>
     <cylinder length="${link3_length}" radius="${link3_width/2}"/>
    </geometry>
    <material name="green"/>
   </visual>
  </link>
  
  <joint name="joint_four" type="revolute">
   <parent link="link_3"/>
   <child link="link_4"/>
   <axis xyz="1 0 0"/>
   <origin xyz="0 0 ${link3_length+0.025}" rpy="0 0 0"/>
   <limit effort="1000.0" lower="0" upper="-1.5" velocity="0.5"/>
  </joint>
  
  <link name="link_4">
   <visual>
    <origin xyz="0 0 ${link4_length/2-0.025}" rpy="0 0 0"/>
    <geometry>
     <cylinder length="${link4_length}" radius="${link4_width/2}"/>
    </geometry>
    <material name="red"/>
   </visual>
  </link>
  
  <joint name="joint_five" type="continuous">
   <parent link="link_4"/>
   <child link="link_5"/>
   <axis xyz="0 0 1"/>
  </joint>
  
  <link name="link_5">
   <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
     <cylinder length="0" radius="0"/>
    </geometry>
    <material name="green"/>
   </visual>
  </link>
  
  <link name="grapper_small">
   <visual>
    <origin xyz="0 0 ${link4_length+0.01}" rpy="0 0 0"/>
    <geometry>
     <box size="0.015 0.01 0.07"/>
    </geometry>
    <material name="blue"/>
   </visual>
  </link>
  
  <link name="grapper_big">
   <visual>
    <origin xyz="0 0 ${link4_length+0.01}" rpy="0 0 0"/>
    <geometry>
     <box size="0.03 0.02 0.07"/>
    </geometry>
    <material name="blue"/>
   </visual>
  </link>
  
  <joint name="base_joint" type="fixed">
    <parent link="base_link"/>
    <child link="link_0"/>
    <origin xyz="0 0 0"/>
  </joint>
  
  <joint name="plack_on_stand" type="fixed">
   <parent link="link_0"/>
   <child link="plack"/>
   <origin xyz="0 0.075 0.025"/>
  </joint>
  
  <joint name="grapper_joint" type="fixed">
    <parent link="link_5"/>
    <child link="grapper_small"/>
    <origin xyz="-0.02 0 0"/>
  </joint>
  
  <joint name="grapper_joint2" type="fixed">
    <parent link="link_5"/>
    <child link="grapper_big"/>
    <origin xyz="0.02 0 0"/>
  </joint>
  
</robot>

